src/Workleap.Extensions.Mongo/Telemetry/DiagnosticsTracingEventSubscriber.cs

```diff
@@ -0,0 +1,152 @@
+
+    #region connection pool events
+    private void Handle(ConnectionPoolAddedConnectionEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolAddingConnectionEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolCheckedInConnectionEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolCheckedOutConnectionEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolCheckingInConnectionEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolCheckingOutConnectionEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolCheckingOutConnectionFailedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolClearedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolClearingEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolClosedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolClosingEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolOpenedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolOpeningEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolReadyEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolRemovedConnectionEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolRemovingConnectionEvent @event) => this.HandleDiagnosticEvent(@event);
+    #endregion
+
+    #region server events
+    private void Handle(ServerClosedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ServerClosingEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ServerDescriptionChangedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ServerHeartbeatFailedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ServerHeartbeatStartedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ServerHeartbeatSucceededEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ServerOpenedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ServerOpeningEvent @event) => this.HandleDiagnosticEvent(@event);
+    #endregion
+
+    private void HandleDiagnosticEvent(object diagnosticEvent)
+    {
+        if (!this._options.Telemetry.CaptureDiagnosticEvents || Activity.Current is null)
+        {
+            return;
+        }
+
+        Type t = diagnosticEvent.GetType();
+        string activityEventName = t.Name;
+        PropertyInfo[] props = t.GetProperties(BindingFlags.Public | BindingFlags.Instance);
```

Comment:
Could it be worth to cache these property infos in a concurrentDictionary? 

Similar to what we do here: https://github.com/gsoft-inc/wl-domain-event-propagation/blob/085f62df83302d55f2cdd25527bd2fdaef598ce4/src/Shared/BaseEventHandler.cs#L36C23-L36C24

Could help save on reflection cost

Reply:
Doing some benchmarks right now

Reply:
![image](https://github.com/gsoft-inc/wl-extensions-mongo/assets/14242083/608d24f1-2681-4c32-9ff8-1a7804cb0832)

On .NET 8, calling Type.GetProperties(...) for ConnectionPoolCheckedOutConnectionEvent (randomly picked this event) costs ~45ns more without caching the properties in a concurrent dictionary (on my machine).

Of course you only get the allocs once per event type

Reply:
@jmcdade11 could you introduce a `ConcurrentDictionary<Type, PropertyInfo[]>` to cache the properties per event type and use `cache.GetOrAdd(eventType, x => x.GetProperties(BindingFlags.Public | BindingFlags.Instance))`?

Reply:
Done